### ISSUE TITLE
Delete the tweet if the message is unpinned

### ISSUE THREAD

**Issue by jotok** [HurrSlack/pinwheel#3]
Ctrl-z

**Comment by zetlen** [comment]
Need a place to store the relation. Redis?

**Comment by zetlen** [comment]
(I checked to see if there were metadata bags on either the tweet object or the pinned post object that we could secretly leverage, but nope)

**Comment by jotok** [comment]
Redis is in-memory only, right? We'll have to see if it would persist between deployments to EB.

**Comment by zetlen** [comment]
[Redis bounces to disk on command](http://redis.io/topics/persistence), which you could set up after a SIGTERM or whatever.

**Comment by zetlen** [comment]
OK, that's not exactly right--RDB is an interval type of thing, which would risk data loss on deploy. Maybe we need just two dumb columns of sqlite or whatever.

**Comment by jotok** [comment]
At the risk of further tying us to AWS, this seems like a good fit for dynamo. We can hide the dynamo-specific calls inside a module to make it easy to swap out.

**Comment by zetlen** [comment]
Sounds good to me. If you can write the implementation I can abstract it into an injectable persistence API

**Comment by jotok** [comment]
All we need is simple key-value storage (and I'm in favor of keeping thing simple). In addition to storing app-relevant data about messages (i.e., tweet ID), I think we may also want to store data for users. For example, we may eventually want to store fake slack user names for each user that we substitute into messages before tweeting, or we may want to allow a user to opt out of tweeting altogether. Rather than create multiple tables, I think we can just a simple convention for converting slack object IDs into hash keys. So, for example, the slack message 
```javascript
{ channel_id: 'ABCDE', ts: 1234.567' }
```
can map to the hash key
```javascript
'm/ABCDE/1234.567'
```
and the user
```javascript
{ id: 'ZYXWV' }
```
can map to the key
```javascript
'u/ZYXWV'
```
We can facilitate the mapping inside the storage interface. I'm currently implementing something that looks like this:
```javascript
var storage = new Storage(env, options);
var message = { channel_id: 'ABCDE', ts: '1234.567' };
var user = { id: 'ZYXWV' };
storage.forMessage(message).put({ tweet_id: 'asdfasdf' });
storage.forUser(user).get(function (err, data) {
  if (data.preferences.neverTweet) {
    console.log('oops');
  }
}
```
By taking this "storage = glorified hashtable" approach, it should be easy to swap in another storage implementation as needed.

**Comment by zetlen** [comment]
I like your key convention, but honestly this whole thing smells pretty relational to me. I think that storage should probably produce a fully denormalized record of everything the bot needs to tweet, but honestly I'm terrible at databases and should be quiet.

**Comment by jotok** [comment]
I like the idea of making a more context-aware API call like:
```javascript
storage.getContextForMessage(message);
// => { user: { neverTweet: false, fakeName: 'ABC' }, channel: { neverTweet: false }, message: { tweetId: 'ABCDE' } }
```
Can you think of any tables that we would add to our schema besides user, channel, and message? If it's just those three, I don't think there is much advantage to switching to a relational storage engine versus making 3 calls to a key-value store to construct the result. And there would be the disadvantage of having to maintain a schema. I'd rather push ahead with dynamo, since we can swap out the storage layer later if we need to. What do you think?

**Comment by zetlen** [comment]
We'll also need a table of our custom slackmoji, but I guess that's it. It's true that it would be a pretty small schema. I'm ok with it!

**Comment by jotok** [comment]
Will the slackmoji table have a foreign key into any other table or vice versa?

**Comment by zetlen** [comment]
Yeah, I don't think so, actually. At runtime, we'd want to look up the full table of slackmoji on every pin and concatenate all their names into a giant alternating regex. If the message matches that regex, it has to get screenshotted for the slackmoji to show up. If not, we can use another static table of slackmoji names to native emoji codepoints, and do a text replace. None of this involves any lookup or joins, so yeah, this is 150 words of irrelevant, isn't it.

**Comment by zetlen** [comment]
Closed by #56
